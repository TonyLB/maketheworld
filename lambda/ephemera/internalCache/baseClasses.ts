import { EphemeraCharacterId, EphemeraRoomId, LegalCharacterColor } from "@tonylb/mtw-interfaces/dist/baseClasses";

export class Deferred <T>{
    invalidationCounter: number;
    promise: Promise<T>;
    _resolve: (value: T) => void = () => {}
    _reject: () => void = () => {}
    isFulfilled: boolean = false
    isRejected: boolean = false
    constructor() {
        this.invalidationCounter = 0
        this.promise = new Promise((resolve, reject)=> {
            this._reject = reject
            this._resolve = resolve
        })
    }

    invalidate() {
        this.invalidationCounter += 1
        if (this.isFulfilled || this.isRejected) {
            this.promise = new Promise((resolve, reject)=> {
                this._reject = reject
                this._resolve = resolve
            })
            this.isFulfilled = false
            this.isRejected = false
        }
    }

    resolve (invalidationCounter: number, value: T): boolean {
        if (invalidationCounter < this.invalidationCounter) {
            return false
        }
        this._resolve(value)
        this.isFulfilled = true
        return true
    }

    reject (invalidationCounter: number): void {
        if (invalidationCounter >= this.invalidationCounter) {
            this._reject()
            this.isRejected = true
        }
    }
}

export type RoomCharacterListItem = {
    EphemeraId: EphemeraCharacterId;
    Color?: LegalCharacterColor;
    ConnectionIds: string[];
    fileURL?: string;
    Name: string;
}

export type DependencyEdge = {
    EphemeraId: string;
    key?: string;
    assets: string[];
}

export type DependencyNode = {
    EphemeraId: string;
    completeness: 'Partial' | 'Complete';
    connections: DependencyEdge[]
}

export type LegalDependencyTag = 'Asset' | 'Variable' | 'Computed' | 'Room' | 'Feature' | 'Map'
export const isLegalDependencyTag = (tag: string): tag is LegalDependencyTag => (['Asset', 'Variable', 'Computed', 'Room', 'Feature', 'Map'].includes(tag))

export type DependencyGraphPut = {
    EphemeraId: string;
    putItem: DependencyEdge;
}

export type DependencyGraphDelete = {
    EphemeraId: string;
    deleteItem: DependencyEdge;
}

export type DependencyGraphAction = DependencyGraphPut | DependencyGraphDelete

export const isDependencyGraphPut = (action: DependencyGraphAction): action is DependencyGraphPut => ('putItem' in action)
export const isDependencyGraphDelete = (action: DependencyGraphAction): action is DependencyGraphDelete => ('deleteItem' in action)

export class CacheBase {
    clear() {}
    async flush() {}
}

type Constructor<T = {}> = new (...args: any[]) => T;

export type CacheConstructor = Constructor<{
    clear(): void;
    flush(): Promise<void>;
}>
