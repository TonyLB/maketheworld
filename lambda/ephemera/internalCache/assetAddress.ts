import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { CacheConstructor } from './baseClasses'
import { AssetWorkspaceAddress } from '@tonylb/mtw-asset-workspace/ts/readOnly'

export type AssetAddressItem = {
    EphemeraId: EphemeraAssetId;
    address: AssetWorkspaceAddress
}

export class CacheAssetAddressData {
    AssetAddressById: Record<EphemeraAssetId, AssetAddressItem> = {};
    clear() {
        this.AssetAddressById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<AssetAddressItem | undefined> {
        return this.AssetAddressById[assetId]
    }
    set(assetItem: AssetAddressItem): void {
        this.AssetAddressById[assetItem.EphemeraId] = assetItem
    }
}

export const CacheAssetAddress = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetAddress extends Base {
        AssetAddress: CacheAssetAddressData = new CacheAssetAddressData()

        override clear() {
            this.AssetAddress.clear()
            super.clear()
        }
    }
}

export default CacheAssetAddress
