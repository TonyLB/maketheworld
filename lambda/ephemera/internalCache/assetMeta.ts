import { EphemeraAssetId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetMetaItem = {
    EphemeraId: EphemeraAssetId;
    found: boolean;
}

export class CacheAssetMetaData {
    AssetMetaById: Record<EphemeraAssetId, AssetMetaItem> = {};
    clear() {
        this.AssetMetaById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetMetaItem, 'found'> | undefined> {
        if (!(this.AssetMetaById[assetId])) {
            const assetData = await ephemeraDB.getItem<AssetMetaItem>({
                    Key: {
                        EphemeraId: assetId,
                        DataCategory: 'Meta::Asset'
                    },
                    ProjectionFields: ['EphemeraId'],
                })
            if (assetData) {
                this.AssetMetaById[assetId] = {
                    ...assetData,
                    found: true
                }
            }
            else {
                this.AssetMetaById[assetId] = {
                    EphemeraId: assetId,
                    found: false
                }
            }
        }
        if (this.AssetMetaById[assetId]?.found) {
            const { found, ...rest } = this.AssetMetaById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetMetaItem, 'found'>): void {
        this.AssetMetaById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetMeta = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetMeta extends Base {
        AssetMeta: CacheAssetMetaData = new CacheAssetMetaData()

        override clear() {
            this.AssetMeta.clear()
            super.clear()
        }
    }
}

export default CacheAssetMeta
