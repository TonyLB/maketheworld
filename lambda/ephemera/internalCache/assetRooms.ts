import { EphemeraAssetId, EphemeraRoomId } from '@tonylb/mtw-interfaces/dist/baseClasses'
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export type AssetRoomsItem = {
    EphemeraId: EphemeraAssetId;
    rooms: EphemeraRoomId[];
    found: boolean;
}

export class CacheAssetRoomsData {
    AssetRoomsById: Record<EphemeraAssetId, AssetRoomsItem> = {};
    clear() {
        this.AssetRoomsById = {}
    }
    async get(assetId: EphemeraAssetId): Promise<Omit<AssetRoomsItem, 'found'> | undefined> {
        if (!(this.AssetRoomsById[assetId])) {
            const assetRooms = await ephemeraDB.query<{ EphemeraId: EphemeraRoomId }[]>({
                IndexName: 'DataCategoryIndex',
                DataCategory: assetId,
                KeyConditionExpression: 'begins_with(EphemeraId, :roomPrefix)',
                ExpressionAttributeValues: {
                    ':roomPrefix': 'ROOM#'
                },        
            })
            if (assetRooms) {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: assetRooms.map(({ EphemeraId }) => (EphemeraId)),
                    found: true
                }
            }
            else {
                this.AssetRoomsById[assetId] = {
                    EphemeraId: assetId,
                    rooms: [],
                    found: false
                }
            }
        }
        if (this.AssetRoomsById[assetId]?.found) {
            const { found, ...rest } = this.AssetRoomsById[assetId]
            return rest
        }
        else {
            return undefined
        }
    }
    set(assetItem: Omit<AssetRoomsItem, 'found'>): void {
        this.AssetRoomsById[assetItem.EphemeraId] = {
            ...assetItem,
            found: true
        }
    }
}

export const CacheAssetRooms = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheAssetRooms extends Base {
        AssetRooms: CacheAssetRoomsData = new CacheAssetRoomsData()

        override clear() {
            this.AssetRooms.clear()
            super.clear()
        }
    }
}

export default CacheAssetRooms
