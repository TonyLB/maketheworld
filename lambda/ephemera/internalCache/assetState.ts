import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

type AssetStateAddress = {
    tag: 'Variable' | 'Computed';
    EphemeraId: string;
}

type AssetStateOutput<T extends Record<string, AssetStateAddress>> = {
    [key in keyof T]: any;
}

export class AssetStateData {
    _StatePromiseByEphemeraId: Record<string, Promise<any>> = {}
    _StateOverrides: Record<string, any> = {}
    
    clear() {
        this._StatePromiseByEphemeraId = {}
    }
    async get<T extends Record<string, AssetStateAddress>>(keys: T): Promise<AssetStateOutput<T>> {
        const itemsInNeedOfFetch = Object.values(keys)
            .filter(({ EphemeraId }) => (!(EphemeraId in this._StatePromiseByEphemeraId)))
        const batchGetPromise = ephemeraDB.batchGetItem<{ EphemeraId: string; value: any; }>({
            Items: itemsInNeedOfFetch.map(({ EphemeraId, tag }) => ({ EphemeraId, DataCategory: `Meta::${tag}` })),
            ProjectionFields: ['EphemeraId', '#value'],
            ExpressionAttributeNames: {
                '#value': 'value'
            }
        })
        Object.values(itemsInNeedOfFetch).forEach(({ EphemeraId }) => {
            this._StatePromiseByEphemeraId[EphemeraId] = batchGetPromise
                .then((items) => (items.find(({ EphemeraId: check }) => (check === EphemeraId))?.value))
                .catch((err) => {
                    if (typeof this._StateOverrides[EphemeraId] === 'undefined') {
                        throw err
                    }
                })
                .then((value) => ((typeof this._StateOverrides[EphemeraId] === 'undefined') ? value : this._StateOverrides[EphemeraId]))
        })
        return Object.assign({}, ...(await Promise.all(
            Object.entries(keys).map(async ([key, { EphemeraId }]) => ({ [key]: await this._StatePromiseByEphemeraId[EphemeraId] }))
        ))) as AssetStateOutput<T>
    }

    set(EphemeraId: string, value: any) {
        this._StateOverrides[EphemeraId] = value
    }

    invalidate(EphemeraId: string) {
        delete this._StateOverrides[EphemeraId]
        delete this._StatePromiseByEphemeraId[EphemeraId]
    }
}

export const AssetState = <GBase extends CacheConstructor>(Base: GBase) => {
    return class AssetState extends Base {
        AssetState: AssetStateData = new AssetStateData()

        override clear() {
            this.AssetState.clear()
            super.clear()
        }
    }
}

export default AssetState
