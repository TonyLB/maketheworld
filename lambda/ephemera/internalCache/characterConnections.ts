import { EphemeraCharacterId } from '@tonylb/mtw-interfaces/ts/baseClasses'
import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { CacheConstructor } from './baseClasses'

export class CacheCharacterConnectionsData {
    CharacterConnectionsById: Record<EphemeraCharacterId, Promise<string[] | undefined>> = {};
    clear() {
        this.CharacterConnectionsById = {}
    }
    async get(characterId: EphemeraCharacterId): Promise<string[] | undefined> {
        if (!(this.CharacterConnectionsById[characterId])) {
            this.CharacterConnectionsById[characterId] = connectionDB.getItem<{ connections: string[] }>({
                    Key: {
                        ConnectionId: characterId,
                        DataCategory: 'Meta::Character'
                    },
                    ProjectionFields: ['connections'],
                }).then((value) => (value?.connections))
        }
        return await this.CharacterConnectionsById[characterId]
    }

    set(characterId: EphemeraCharacterId, connections: string[]): void {
        this.CharacterConnectionsById[characterId] = Promise.resolve(connections)
    }
}

export const CacheCharacterConnections = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheCharacterConnections extends Base {
        CharacterConnections: CacheCharacterConnectionsData = new CacheCharacterConnectionsData()

        override clear() {
            this.CharacterConnections.clear()
            super.clear()
        }
    }
}

export default CacheCharacterConnections
