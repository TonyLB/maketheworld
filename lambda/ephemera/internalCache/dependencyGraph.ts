import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB'
import { deepEqual } from '@tonylb/mtw-utilities/dist/objects';
import { DependencyNodeNonAsset } from '../messageBus/baseClasses';
import { CacheConstructor } from './baseClasses'

type DependencyNodeGraphConnection = {
    EphemeraId: string;
    key?: string;
}

export type DependencyNodeGraphItem = Omit<DependencyNodeNonAsset, 'connections'> & {
    completeness: 'Partial' | 'Complete';
    connections: DependencyNodeGraphConnection[]
}

export class DependencyGraphData {
    _Promises: Record<string, Promise<DependencyNodeGraphItem>> = {}
    _Store: Record<string, DependencyNodeGraphItem> = {}
    
    clear() {
        this._Promises = {}
        this._Store = {}
    }

    getPartial(EphemeraId: string): DependencyNodeNonAsset | undefined {
        if (!(EphemeraId in this._Store)) {
            return undefined
        }
        const currentNode = this._Store[EphemeraId]
        const remappedConnections = currentNode.connections
            .map(({ EphemeraId, key }) => {
                const recurse = this.getPartial(EphemeraId)
                if (typeof recurse === 'undefined') {
                    return undefined
                }
                const returnValue = {
                    ...recurse,
                    key
                } as DependencyNodeNonAsset
                return returnValue
            })
        return {
            ...currentNode,
            connections: remappedConnections
                .filter((value: DependencyNodeNonAsset | undefined): value is DependencyNodeNonAsset => (Boolean(value)))
        }
    }

    put(EphemeraId: string, value: DependencyNodeNonAsset) {

    }

    delete(EphemeraId: string, descendant: string) {
        
    }

    invalidate(EphemeraId: string) {
    }

    isComplete(EphemeraId: string): boolean {
        if (!(EphemeraId in this._Store)) {
            return false
        }
        const currentNode = this._Store[EphemeraId]
        if (currentNode.completeness === 'Partial') {
            return false
        }
        return currentNode.connections.reduce((previous, { EphemeraId }) => (previous && this.isComplete(EphemeraId)), true)
    }

}

export const DependencyGraph = <GBase extends CacheConstructor>(Base: GBase) => {
    return class DependencyGraph extends Base {
        Descent: DependencyGraphData;
        Ancestry: DependencyGraphData;

        constructor(...rest: any) {
            super()
            this.Descent = new DependencyGraphData()
            this.Ancestry = new DependencyGraphData()
        }
        override clear() {
            this.Descent.clear()
            this.Ancestry.clear()
            super.clear()
        }
    }
}

export default DependencyGraph
