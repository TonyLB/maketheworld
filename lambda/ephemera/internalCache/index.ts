import { CacheConstructor, CacheBase } from './baseClasses'

import { connectionDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'
import { delayPromise } from '@tonylb/mtw-utilities/dist/dynamoDB/delayPromise'
import CacheRoomCharacterLists from './roomCharacterLists';
import CacheCharacterMeta from './characterMeta';
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB';
import CacheCharacterConnections from './characterConnections';
import AssetState from './assetState';
import DependencyGraph from './dependencyGraph';
import ComponentMeta from './componentMeta';
import ComponentRender from './componentRender';

type CacheGlobalKeys = 'ConnectionId' | 'RequestId' | 'player' | 'assets' | 'connections'
class CacheGlobalData {
    ConnectionId?: string;
    RequestId?: string;
    player?: string;
    assets?: string[];
    connections?: string[];
    get(key: 'ConnectionId' | 'RequestId' | 'player'): Promise<string | undefined>
    get(key: 'assets' | 'connections'): Promise<string[] | undefined>
    get(key: CacheGlobalKeys): Promise<string | string[] | undefined>
    async get(key: CacheGlobalKeys) {
        switch(key) {
            case 'player':
                if (!this.player && this.ConnectionId) {
                    //
                    // TODO: Replace repeated attempts with exponential backoff by
                    // refactoring connectionDB.getItem to allow a consistent argument
                    // that can actviate strongly-consistent reads
                    //
                    let attempts = 0
                    let exponentialBackoff = 50
                    while(attempts < 5) {
                        const { player = '' } = await connectionDB.getItem<{ player: string }>({
                            ConnectionId: `CONNECTION#${this.ConnectionId}`,
                            DataCategory: 'Meta::Connection',
                            ProjectionFields: ['player']
                        }) || {}
                        if (player) {
                            this.player = player
                            return player
                        }
                        attempts += 1
                        await delayPromise(exponentialBackoff)
                        exponentialBackoff = exponentialBackoff * 2
                    }
                    console.log(`Exponential backoff on player caching failed after five attempts (${this.ConnectionId})`)
                }
                return this.player
            case 'assets':
                if (typeof this.assets === 'undefined') {
                    const { assets = [] } = (await ephemeraDB.getItem<{ assets: string[] }>({
                        EphemeraId: 'Global',
                        DataCategory: 'Assets',
                        ProjectionFields: ['assets']
                    })) || {}
                    this.assets = assets
                }
                return this.assets
            case 'connections':
                if (typeof this.connections === 'undefined') {
                    const { connections = {} } = (await connectionDB.getItem<{ connections: Record<string, string> }>({
                        ConnectionId: 'Global',
                        DataCategory: 'Connections',
                        ProjectionFields: ['connections']
                    })) || {}
                    this.connections = Object.keys(connections)
                }
                return this.connections
            default:
                return this[key]
        }
    }

    clear() {
        this.ConnectionId = undefined
        this.RequestId = undefined
        this.player = undefined
        this.assets = undefined
    }

    set(props: { key: 'ConnectionId' | 'RequestId', value: string; }): void {
        this[props.key] = props.value
    }
}

export const CacheGlobal = <GBase extends CacheConstructor>(Base: GBase) => {
    return class CacheGlobal extends Base {
        Global: CacheGlobalData = new CacheGlobalData()

        override clear() {
            this.Global.clear()
            super.clear()
        }
    }
}

const InternalCache = ComponentRender(ComponentMeta(AssetState(DependencyGraph(CacheCharacterConnections(CacheCharacterMeta(CacheRoomCharacterLists(CacheGlobal(CacheBase))))))))
export const internalCache = new InternalCache()
export default internalCache
