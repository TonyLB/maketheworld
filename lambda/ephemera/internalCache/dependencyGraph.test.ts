jest.mock('@tonylb/mtw-utilities/dist/dynamoDB/index')
import { ephemeraDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'

import internalCache from "."
import { AssetStateMapping } from './assetState'
import { DependencyNodeGraphItem } from './dependencyGraph'

const ephemeraMock = ephemeraDB as jest.Mocked<typeof ephemeraDB>

describe('DependencyGraph', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        internalCache.clear()
    })

    const testStore: Record<string, DependencyNodeGraphItem> = {
        testOne: {
            EphemeraId: 'testOne',
            tag: 'Variable',
            completeness: 'Complete',
            assets: ['base'],
            connections: [
                { EphemeraId: 'testTwo', key: 'testOne' },
                { EphemeraId: 'testThree', key: 'testOne' }
            ]
},
        testTwo: {
            EphemeraId: 'testTwo',
            tag: 'Computed',
            completeness: 'Partial',
            assets: ['base'],
            connections: [{ EphemeraId: 'testThree', key: 'testTwo' }]
        },
        testThree: {
            EphemeraId: 'testThree',
            tag: 'Computed',
            completeness: 'Partial',
            assets: ['base'],
            connections: []
        }
    }

    describe('getPartial', () => {
        it('should correctly decipher a partial tree', () => {
            internalCache.Descent._Store = testStore
    
            expect(internalCache.Descent.getPartial('testOne')).toMatchSnapshot()
        })
    })

    describe('isComplete', () => {
        it('should correctly tag a partial tree', () => {
            internalCache.Descent._Store = testStore
    
            expect(internalCache.Descent.isComplete('testOne')).toBe(false)
        })

        it('should correctly flag a complete tree', () => {
            internalCache.Descent._Store = {
                testOne: {
                    EphemeraId: 'testOne',
                    tag: 'Variable',
                    completeness: 'Complete',
                    assets: ['base'],
                    connections: [
                        { EphemeraId: 'testTwo', key: 'testOne' },
                        { EphemeraId: 'testThree', key: 'testOne' }
                    ]
                },
                testTwo: {
                    EphemeraId: 'testTwo',
                    tag: 'Computed',
                    completeness: 'Complete',
                    assets: ['base'],
                    connections: [{ EphemeraId: 'testThree', key: 'testTwo' }]
                },
                testThree: {
                    EphemeraId: 'testThree',
                    tag: 'Computed',
                    completeness: 'Complete',
                    assets: ['base'],
                    connections: []
                }
            }
    
            expect(internalCache.Descent.isComplete('testOne')).toBe(true)
        })
    })

    describe('put', () => {
        it('should correctly add a single graph item', () => {
            internalCache.Descent._Store = testStore
            internalCache.Descent.put({
                EphemeraId: 'testFour',
                tag: 'Variable',
                completeness: 'Complete',
                assets: ['base'],
                connections: [{ EphemeraId: 'testTwo', key: 'testFour' }]
            })
            expect(internalCache.Descent.getPartial('testFour')).toMatchSnapshot()
        })

        it('should correctly add a complete tree', () => {
            internalCache.Descent._Store = testStore
            internalCache.Descent.put({
                EphemeraId: 'testFour',
                tag: 'Variable',
                assets: ['base'],
                connections: [
                    {
                        EphemeraId: 'testTwo',
                        tag: 'Computed',
                        key: 'testFour',
                        assets: ['base'],
                        connections: [{
                            EphemeraId: 'testThree',
                            key: 'testTwo',
                            assets: ['base'],
                            connections: [],
                            tag: 'Computed'
                        }]
                    }
                ]
            })
            expect(internalCache.Descent.getPartial('testFour')).toMatchSnapshot()
        })
    })

    describe('delete', () => {
        it('should correctly delete a connection', () => {
            internalCache.Descent._Store = testStore
            internalCache.Descent.delete('testTwo', 'testThree')
            expect(internalCache.Descent.getPartial('testOne')).toMatchSnapshot()
        })
    })
})