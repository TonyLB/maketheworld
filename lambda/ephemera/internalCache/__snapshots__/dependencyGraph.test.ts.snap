// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyGraph delete should correctly delete a connection 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
      },
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`DependencyGraph delete should decrement connection assets when redundancies exist 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
          "layer",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`DependencyGraph get should correctly fetch a tree 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
    "tag": "Computed",
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "connections": Array [],
    "tag": "Computed",
  },
]
`;

exports[`DependencyGraph getPartial should correctly decipher a partial tree 1`] = `
Array [
  Object {
    "EphemeraId": "VARIABLE#testOne",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
      },
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`DependencyGraph put should backpopulate antidependency links 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "assets": Array [],
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testFour",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
    "tag": "Computed",
  },
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;

exports[`DependencyGraph put should correctly add a complete tree 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Complete",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Complete",
    "connections": Array [],
  },
]
`;

exports[`DependencyGraph put should correctly add a single graph item 1`] = `
Array [
  Object {
    "EphemeraId": "COMPUTED#testFour",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testTwo",
        "assets": Array [
          "base",
        ],
        "key": "testFour",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testTwo",
    "completeness": "Partial",
    "connections": Array [
      Object {
        "EphemeraId": "COMPUTED#testThree",
        "assets": Array [
          "base",
        ],
        "key": "testTwo",
      },
    ],
  },
  Object {
    "EphemeraId": "COMPUTED#testThree",
    "completeness": "Partial",
    "connections": Array [],
  },
]
`;
